### PR TITLE
Port Hibernate's support subqueries in HQL as CASE statement alternatives

### PR DESCRIPTION
**Pull request by maca88**
Port [HHH-1689](https://github.com/hibernate/hibernate-orm/commit/57401b9aa804c59bcad765240971f2aa0a1a2f3a#diff-4ba4b0e3f3528d1a0d265471f19bf43d)

This feature is needed for #2079 which converts the whole conditional expression to sql. Tests for linq will be added with #2079 as currently it is not possible to test it due to the split of the conditional expression into separate columns which are then evaluated on the client side.

### CODE REVIEW

**Comment by maca88** [comment]
As we have a failing test by just porting the feature as it is, I've modified the case expression definition to be consistent with [Sql Server CASE](https://docs.microsoft.com/en-us/sql/t-sql/language-elements/case-transact-sql?view=sql-server-2017)  definition and added an additional test for simple case expression.

File: src/NHibernate.Test/Hql/SubQueryTest.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+		}
+
+		[TestCase("SELECT CASE WHEN l.id IS NOT NULL THEN (SELECT COUNT(r.id) FROM Root r) ELSE 0 END FROM Leaf l")]
+		[TestCase("SELECT CASE WHEN (SELECT COUNT(r.id) FROM Root r) > 1 THEN 1 ELSE 0 END FROM Leaf l")]
+		[TestCase("SELECT CASE WHEN  l.id > 1 THEN 1 ELSE (SELECT COUNT(r.id) FROM Root r) END FROM Leaf l")]
+		[TestCase("SELECT CASE (SELECT COUNT(r.id) FROM Root r) WHEN  1 THEN 1 ELSE 0 END FROM Leaf l")]
+		[TestCase("SELECT CASE l.id WHEN (SELECT COUNT(r.id) FROM Root r) THEN 1 ELSE 0 END FROM Leaf l")]
+		public void TestSubQuery(string query)
+		{
+			if (!Dialect.SupportsScalarSubSelects)
```
Comment (hazzik) [inline src/NHibernate.Test/Hql/SubQueryTest.cs]: Should be in `AppliesTo`

File: src/NHibernate/Hql/Ast/ANTLR/Hql.g [hunk 2]
```diff
@@ -527,23 +527,32 @@ unaryExpression
 	;
 	
 caseExpression
-	: CASE (whenClause)+ (elseClause)? END
-		-> ^(CASE whenClause+ elseClause?) 
-	| CASE unaryExpression (altWhenClause)+ (elseClause)? END
-		-> ^(CASE2 unaryExpression altWhenClause+ elseClause?)
+	// NOTE : the unaryExpression rule contains the subQuery rule
```
Comment (hazzik) [inline src/NHibernate/Hql/Ast/ANTLR/Hql.g]: Is this note relevant?

Comment (maca88) [reply src/NHibernate/Hql/Ast/ANTLR/Hql.g]: Not anymore, removed.

File: src/NHibernate.Test/Hql/SubQueryTest.cs (hunk 1 above)
Comment (maca88) [reply src/NHibernate.Test/Hql/SubQueryTest.cs]: Done.

**Pull request merged**